File: src/main/org/nlogo/hubnet/mirroring/ClientWorldS.java
package org.nlogo.hubnet.mirroring;

import java.util.Comparator;

class ClientWorldS {

  // this class is not instantiable
  private ClientWorldS() {
    throw new IllegalStateException();
  }

  static class TurtleKey {
    long who;
    int breedIndex;

    public TurtleKey(long who, int breedIndex) {
      this.who = who;
      this.breedIndex = breedIndex;
    }

    @Override
    public boolean equals(Object o) {
      return (who == ((TurtleKey) o).who) &&
          (breedIndex == ((TurtleKey) o).breedIndex);
    }

    @Override
    public int hashCode() {
      return (int) (breedIndex * 1000 + who);
    }

    @Override
    public String toString() {
      return "(" + who + " " + breedIndex + ")";
    }
  }

  static class TurtleKeyComparator
      implements Comparator<TurtleKey> {
    public int compare(TurtleKey tk1, TurtleKey tk2) {
      if (tk1.breedIndex == tk2.breedIndex) {
        return (int) (tk1.who - tk2.who);
      } else {
        return tk1.breedIndex - tk2.breedIndex;
      }
    }
  }

  static class LinkKey {
    long id;
    long end1;
    long end2;
    int breedIndex;

    public LinkKey(long id, long end1, long end2, int breedIndex) {
      this.id = id;
      this.end1 = end1;
      this.end2 = end2;
      this.breedIndex = breedIndex;
    }

    @Override
    public boolean equals(Object o) {
      return id == ((LinkKey) o).id;
    }

    @Override
    public int hashCode() {
      return (int) id;
    }

    @Override
    public String toString() {
      return "(" + id + " " + breedIndex + ")";
    }
  }

  static class LinkKeyComparator
      implements Comparator<LinkKey> {
    public int compare(LinkKey key1, LinkKey key2) {
      if (key1.end1 == key2.end1) {
        if (key1.end2 == key2.end2) {
          if (key1.breedIndex == key2.breedIndex) {
            return (int) (key1.id - key2.id);
          } else {
            return key1.breedIndex - key2.breedIndex;
          }
        } else {
          return (int) (key1.end2 - key2.end2);
        }
      } else {
        return (int) (key1.end1 - key2.end1);
      }
    }
  }

}

File: src/main/org/nlogo/hubnet/mirroring/ClientWorldS.scala
package org.nlogo.hubnet.mirroring

import java.util.Comparator

private object ClientWorldS {

  case class TurtleKey(who: Long, breedIndex: Int)

  class TurtleKeyComparator extends Comparator[TurtleKey] {
    override def compare(tk1: TurtleKey, tk2: TurtleKey) =
      if (tk1.breedIndex == tk2.breedIndex)
        (tk1.who - tk2.who).toInt
      else
        tk1.breedIndex - tk2.breedIndex
  }

  case class LinkKey(id: Long, end1: Long, end2: Long, breedIndex: Int)

  class LinkKeyComparator extends Comparator[LinkKey] {
    override def compare(key1: LinkKey, key2: LinkKey): Int = {
      if (key1.end1 == key2.end1) {
        if (key1.end2 == key2.end2) {
          if (key1.breedIndex == key2.breedIndex) {
            return (key1.id - key2.id).toInt
          } else {
            return key1.breedIndex - key2.breedIndex
          }
        } else {
          return (key1.end2 - key2.end2).toInt
        }
      } else {
        return (key1.end1 - key2.end1).toInt
      }
    }
  }

}
